lib/rucio/db/sqla/__init__.py

```diff
@@ -3,3 +3,27 @@
 # Licensed under the Apache License, Version 2.0 (the "License");
 # You may not use this file except in compliance with the License.
 # You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+from sqlalchemy.sql.expression import bindparam, text
+
+
+def filter_thread_work(session, query, total_threads, thread, hash_variable=None):
+    """ Filters a query to partition thread workloads based on the thread id and total number of threads """
+    if thread and total_threads and (total_threads-1) > 0:
+        if session.bind.dialect.name == 'oracle':
+            bindparams = [bindparam('thread_number', thread), bindparam('total_threads', total_threads)]
+            if not hash_variable:
+                query = query.filter(text('ORA_HASH(id, :total_threads) = :thread_number', bindparams=bindparams))
+            else:
+                query = query.filter(text('ORA_HASH(%s, :total_workers) = :worker_number' % (hash_variable), bindparams=bindparams))
+        elif session.bind.dialect.name == 'mysql':
+            if not hash_variable:
+                query = query.filter(text('mod(md5(id), %s) = %s' % (total_threads, thread)))
+            else:
+                query = query.filter(text('mod(md5(%s), %s) = %s' % (hash_variable, total_workers, worker_number)))
+        elif session.bind.dialect.name == 'postgresql':
+            if not hash_variable:
+                query = query.filter(text('mod(abs((\'x\'||md5(id::text))::bit(32)::int), %s) = %s' % (total_threads, thread)))
+            else:
+                query = query.filter(text('mod(abs((\'x\'||md5(%s::text))::bit(32)::int), %s) = %s' % (hash_variable, total_workers, worker_number)))
+    return query
```

Comment:
Hi @bjwhite-fnal 
This looks very good to me. Instead of `__init__` I would add this to the `sqla/util.py` though.

In the other files I see there was big confusion between `total_workers + 1` and `total_workers -1` etc.
There is one remaining issue which would be great if you can check. In the daemons calling these core functions you changed we usually do `total_workers=heartbeat['nr_threads'] - 1` and `worker_number=heartbeat['assign_thread']` - Please check if this is done in all daemons consistently or if we have some +/-1 mixup there as well.

Reply:
My first thought was also to include this definition in `util.py`. I tried that first, however it created a cyclic importing issue with another file. `util.py` wanted to import a module that ended up using `filter_thread_work()`, which had not been defined yet. There are no forward declarations in Python, so I moved it to `__init__.py` as opposed to creating a new module.

I'll check on the rest of that for you.

Reply:
My first thought was also to include this definition in util.py. I tried that first, however it created a cyclic importing issue with another file. util.py wanted to import a module that ended up using filter_thread_work(), which had not been defined yet. There are no forward declarations in Python, so I moved it to __init__.py as opposed to creating a new module.

I'll check on the rest of that for you.